File: src/itkach/aard2/article/ArticleCollectionActivity.java
package itkach.aard2.article;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Intent;
import android.content.SharedPreferences;
import android.database.DataSetObserver;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.util.TypedValue;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.view.ViewGroup;
import android.widget.Toast;

import android.content.Context;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NavUtils;
import androidx.core.app.TaskStackBuilder;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.core.view.WindowInsetsControllerCompat;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager.widget.PagerTitleStrip;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.color.MaterialColors;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import java.util.List;
import java.util.Objects;

import itkach.aard2.BuildConfig;
import itkach.aard2.MainActivity;
import itkach.aard2.R;
import itkach.aard2.SlobHelper;
import itkach.aard2.prefs.AppPrefs;
import itkach.aard2.prefs.ArticleCollectionPrefs;
import itkach.aard2.slob.SlobTags;
import itkach.aard2.utils.ThreadUtils;
import itkach.aard2.utils.Utils;
import itkach.aard2.widget.ArticleWebView;
import itkach.slob.Slob;

public class ArticleCollectionActivity extends AppCompatActivity
        implements SharedPreferences.OnSharedPreferenceChangeListener {
    public static final String ACTION_BOOKMARKS = BuildConfig.APPLICATION_ID + ".action.BOOKMARKS";
    public static final String ACTION_HISTORY = BuildConfig.APPLICATION_ID + ".action.HISTORY";

    private static final String TAG = ArticleCollectionActivity.class.getSimpleName();

    private ArticleCollectionPagerAdapter pagerAdapter;
    private ViewPager2 viewPager;
    private ArticleCollectionViewModel viewModel;
    private boolean isHistory;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        EdgeToEdge.enable(this);
        super.onCreate(savedInstanceState);
        Utils.updateNightMode();
        setContentView(R.layout.activity_article_collection_loading);
        setSupportActionBar(findViewById(R.id.toolbar));
        viewModel = new ViewModelProvider(this).get(ArticleCollectionViewModel.class);

        final ActionBar actionBar = requireActionBar();
        actionBar.hide();
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setSubtitle("...");

        final Intent intent = getIntent();
        final int position = intent.getIntExtra("position", 0);
        isHistory = ArticleCollectionActivity.ACTION_HISTORY.equals(intent.getAction());

        viewModel.getBlobListLiveData().observe(this, blobListWrapper -> {
            if (blobListWrapper == null) {
                // Adapter is never null, empty or less than |position|
                return;
            }

            setContentView(R.layout.activity_article_collection);
            setSupportActionBar(findViewById(R.id.toolbar));
            requireActionBar().setDisplayHomeAsUpEnabled(true);

            TabLayout tabs = findViewById(R.id.tabs);
            tabs.setVisibility(
                    blobListWrapper.size() == 1 ? ViewGroup.GONE : ViewGroup.VISIBLE);

            viewPager = findViewById(R.id.pager);
            viewPager.setNestedScrollingEnabled(true);
            pagerAdapter = new ArticleCollectionPagerAdapter(blobListWrapper, this);
            viewPager.setAdapter(pagerAdapter);
            TabLayoutMediator tabLayoutMediator = new TabLayoutMediator(tabs, viewPager, true, (tab, position1) -> {
                if (!ArticleCollectionPrefs.disableTabLabels()) {
                    tab.setText(pagerAdapter.getPageTitle(position1));
                }
            });
            tabLayoutMediator.attach();
            viewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
                @Override
                public void onPageScrollStateChanged(int arg0) {
                }

                @Override
                public void onPageScrolled(int arg0, float arg1, int arg2) {
                }

                @Override
                public void onPageSelected(final int position) {
                    updateTitle(position);
                    ArticleFragment fragment = pagerAdapter.getPageFragment(position);
                    if (fragment != null) {
                        pagerAdapter.setPrimaryItem(fragment);
                        runOnUiThread(() -> {
                            fragment.applyTextZoomPref();
                        });
                    }

                }
            });
//            viewPager.setCurrentItem(position);
//            updateTitle(position);
            pagerAdapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
                @Override
                public void onChanged() {
                    if (pagerAdapter.getItemCount() == 0) {
                        finish();
                    }
                }
            });
            viewPager.setCurrentItem(position, false);
            updateTitle(position);

            ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.layout), (v, insets) -> {
                Insets bars = insets.getInsets(
                        WindowInsetsCompat.Type.systemBars()
                                | WindowInsetsCompat.Type.displayCutout()
                );
                ViewGroup.MarginLayoutParams mlp = (ViewGroup.MarginLayoutParams) v.getLayoutParams();
                mlp.leftMargin = bars.left;
                mlp.bottomMargin = (ArticleCollectionPrefs.isFullscreen()) ? 0 : bars.bottom;
                mlp.topMargin = (ArticleCollectionPrefs.isFullscreen()) ? 0 : bars.top;
                mlp.rightMargin = bars.right;
                v.setLayoutParams(mlp);
                return WindowInsetsCompat.CONSUMED;
            });
        });
        viewModel.getFailureMessageLiveData().observe(this, message -> {
            if (message != null) {
                new Handler(Looper.getMainLooper()).post(() -> {
                    if (AppPrefs.openMissingInBrowser() && viewModel.articleUri != null) {
                        openUrlInBrowser(viewModel.articleUri);
                    } else {
                        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
                    }
                    finish();
                });
            }
        });
        // Load adapter
        viewModel.loadBlobList(intent);

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.layout), (v, insets) -> {
            Insets bars = insets.getInsets(
                    WindowInsetsCompat.Type.systemBars()
            );
            ViewGroup.MarginLayoutParams mlp = (ViewGroup.MarginLayoutParams) v.getLayoutParams();
            mlp.leftMargin = bars.left;
            mlp.bottomMargin = (ArticleCollectionPrefs.isFullscreen()) ? 0 : bars.bottom;
            mlp.topMargin = bars.top;
            mlp.rightMargin = bars.right;
            v.setLayoutParams(mlp);
            return WindowInsetsCompat.CONSUMED;
        });
    }

    @Override
    public void onBackPressed() {
        if (ArticleCollectionPrefs.isFullscreen()) {
            // Exit fullscreen
            toggleFullScreen();
            return;
        }
        super.onBackPressed();
    }

    @NonNull
    public ActionBar requireActionBar() {
        return Objects.requireNonNull(getSupportActionBar());
    }

    private void updateTitle(int position) {
        Slob.Blob blob = pagerAdapter.get(position);
        CharSequence pageTitle = pagerAdapter.getPageTitle(position);
        ActionBar actionBar = requireActionBar();
        if (blob != null) {
            String dictLabel = blob.owner.getTags().get(SlobTags.TAG_LABEL);
            actionBar.setTitle(dictLabel);
            if (!AppPrefs.disableHistory() && !isHistory) {
                SlobHelper slobHelper = SlobHelper.getInstance();
                slobHelper.history.add(slobHelper.getHttpUri(blob));
            }
        } else {
            actionBar.setTitle("???");
        }
        actionBar.setSubtitle(pageTitle);
    }


    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (key.equals(ArticleCollectionPrefs.PREF_FULLSCREEN)) {
            applyFullScreenPref();
        }
    }

    private void applyFullScreenPref() {
        if (ArticleCollectionPrefs.isFullscreen()) {
            WindowInsetsControllerCompat windowInsetsController = WindowCompat.getInsetsController(getWindow(),
                    getWindow().getDecorView());
            windowInsetsController.hide(WindowInsetsCompat.Type.systemBars());
            windowInsetsController.setSystemBarsBehavior(WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE);
            requireActionBar().hide();
        } else {
            WindowInsetsControllerCompat windowInsetsController = WindowCompat.getInsetsController(getWindow(),
                    getWindow().getDecorView());
            windowInsetsController.show(WindowInsetsCompat.Type.systemBars());
            requireActionBar().show();
        }
    }

    SharedPreferences prefs() {
        return getSharedPreferences("articleCollection", Activity.MODE_PRIVATE);
    }

    void toggleFullScreen() {
        ArticleCollectionPrefs.setFullscreen(!ArticleCollectionPrefs.isFullscreen());
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.d(TAG, "[F] Resume");
        applyFullScreenPref();
        prefs().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.d(TAG, "[F] Pause");
        prefs().unregisterOnSharedPreferenceChangeListener(this);
    }

    @Override
    protected void onDestroy() {
        if (viewPager != null) {
            viewPager.setAdapter(null);
        }
        if (pagerAdapter != null) {
            pagerAdapter.destroy();
        }
        super.onDestroy();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            Intent upIntent = Intent.makeMainActivity(new ComponentName(this, MainActivity.class));
            if (NavUtils.shouldUpRecreateTask(this, upIntent)) {
                TaskStackBuilder.create(this)
                        .addNextIntent(upIntent).startActivities();
                finish();
            } else {
                // This activity is part of the application's task, so simply
                // navigate up to the hierarchical parent activity.
                upIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
                startActivity(upIntent);
                finish();
            }
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onKeyUp(int keyCode, @NonNull KeyEvent event) {
        if (event.isCanceled()) {
            return true;
        }
        if (pagerAdapter == null) {
            return false;
        }
        ArticleFragment af = pagerAdapter.getPrimaryItem();
        if (af != null) {
            ArticleWebView webView = af.getWebView();
            if (webView != null) {
                if (keyCode == KeyEvent.KEYCODE_BACK) {
                    if (webView.canGoBack()) {
                        webView.goBack();
                        return true;
                    }
                }

                if (keyCode == KeyEvent.KEYCODE_VOLUME_UP) {
                    if (!AppPrefs.useVolumeKeysForNavigation()) {
                        return false;
                    }
                    boolean scrolled = webView.pageUp(false);
                    if (!scrolled) {
                        int current = viewPager.getCurrentItem();
                        if (current > 0) {
                            viewPager.setCurrentItem(current - 1);
                        } else {
                            finish();
                        }
                    }
                    return true;
                }
                if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
                    if (!AppPrefs.useVolumeKeysForNavigation()) {
                        return false;
                    }
                    boolean scrolled = webView.pageDown(false);
                    if (!scrolled) {
                        int current = viewPager.getCurrentItem();
                        if (current < pagerAdapter.getItemCount() - 1) {
                            viewPager.setCurrentItem(current + 1);
                        }
                    }
                    return true;
                }
            }
        }
        return super.onKeyUp(keyCode, event);
    }

    @Override
    public boolean onKeyDown(int keyCode, @NonNull KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_VOLUME_UP || keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
            if (!AppPrefs.useVolumeKeysForNavigation()) {
                return false;
            }
            event.startTracking();
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public boolean onKeyLongPress(int keyCode, KeyEvent event) {
        if (!AppPrefs.useVolumeKeysForNavigation()) {
            return false;
        }
        ArticleFragment af = pagerAdapter.getPrimaryItem();
        if (af != null) {
            ArticleWebView webView = af.getWebView();

            if (keyCode == KeyEvent.KEYCODE_VOLUME_UP) {
                webView.pageUp(true);
                return true;
            }
            if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
                webView.pageDown(true);
                return true;
            }
        }
        return super.onKeyLongPress(keyCode, event);
    }

    public static class ArticleCollectionPagerAdapter extends FragmentStateAdapter {
        private final BlobListWrapper blobListWrapper;
        private final FragmentActivity activity;
        private final DataSetObserver observer = new DataSetObserver() {
            @Override
            public void onChanged() {
                ThreadUtils.postOnMainThread(() -> notifyDataSetChanged());
            }
        };

        private ArticleFragment primaryItem;

        public ArticleCollectionPagerAdapter(@NonNull BlobListWrapper blobListWrapper, FragmentActivity activity) {
            super(activity);
            this.blobListWrapper = blobListWrapper;
            this.activity = activity;
            this.blobListWrapper.registerDataSetObserver(observer);
        }

        public void destroy() {
            blobListWrapper.unregisterDataSetObserver(observer);
        }

        public void setPrimaryItem(@NonNull ArticleFragment object) {
            this.primaryItem = (ArticleFragment) object;
        }

        public ArticleFragment getPrimaryItem() {
            return this.primaryItem;
        }

        @Override
        @NonNull
        public ArticleFragment createFragment(int i) {
            ArticleFragment fragment = new ArticleFragment();

            Slob.Blob blob = get(i);
            if (blob != null) {
                Uri articleUri = SlobHelper.getInstance().getHttpUri(blob);
                Bundle args = new Bundle();
                args.putParcelable(ArticleFragment.ARG_URI, articleUri);
                fragment.setArguments(args);
            }
            return fragment;
        }

        @Override
        public int getItemCount() {
            return blobListWrapper.size();
        }

        public Slob.Blob get(int position) {
            return blobListWrapper.get(position);
        }

        @Override
        public long getItemId(int position) {
            // we override getItemId to ensure position is the id
            // so that we can find a fragment based on a position
            return position;
        }

        public ArticleFragment getPageFragment(long id) {
            return (ArticleFragment) this.activity.getSupportFragmentManager().findFragmentByTag("f" + Long.toString(id));
        }

        public CharSequence getPageTitle(int position) {
            CharSequence label = blobListWrapper.getLabel(position);
            return label != null ? label : "???";
        }

    }
    
    public void openUrlInBrowser(@NonNull Uri url) {
        Intent intent = new Intent(Intent.ACTION_VIEW, url);
        intent.addCategory(Intent.CATEGORY_BROWSABLE);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        
        // Get all apps that can handle this intent
        PackageManager pm = getPackageManager();
        List<ResolveInfo> resolveInfos = pm.queryIntentActivities(intent, 0);
        
        for (ResolveInfo resolveInfo : resolveInfos) {
            String packageName = resolveInfo.activityInfo.packageName;

            // Skip your own app's package
            if (!packageName.equals(getPackageName())) {
                intent.setPackage(packageName);
                startActivity(intent);
                break;
            }
        }

    }
}

File: src/itkach/aard2/ArticleCollectionActivity.kt
package itkach.aard2

import android.app.SearchManager
import android.content.ComponentName
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.database.DataSetObserver
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.KeyEvent
import android.view.MenuItem
import android.view.View
import android.view.View.OnSystemUiVisibilityChangeListener
import android.view.ViewGroup
import android.view.Window
import android.widget.BaseAdapter
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NavUtils
import androidx.core.app.TaskStackBuilder
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView.AdapterDataObserver
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import itkach.aard2.MainActivity
import itkach.slob.Slob
import itkach.slob.Slob.PeekableIterator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ArticleCollectionActivity : AppCompatActivity(), OnSystemUiVisibilityChangeListener,
    OnSharedPreferenceChangeListener {
    var articleCollectionPagerAdapter: ArticleCollectionPagerAdapter? = null
    private var viewPager: ViewPager2? = null


    internal inner class ToBlobWithFragment(private val fragment: String) : ToBlob {
        override fun convert(item: Any): Slob.Blob? {
            val b = item as Slob.Blob
            return Slob.Blob(b.owner, b.id, b.key, this.fragment)
        }
    }

    private var blobToBlob: ToBlob = object : ToBlob {
        override fun convert(item: Any): Slob.Blob? {
            return item as Slob.Blob
        }
    }


    private var onDestroyCalled = false


    val currentPosition: Int
        get() = if (viewPager != null) viewPager!!.currentItem else -1

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_PROGRESS)
        val app = application as Application
        app.installTheme(this)
        supportActionBar!!.hide()
        setContentView(R.layout.activity_article_collection_loading)
        app.push(this)
        val actionBar = supportActionBar
        actionBar!!.setDisplayHomeAsUpEnabled(true)
        actionBar.setSubtitle("...")
        val intent = intent
        val position = intent.getIntExtra("position", 0)

        val deferred = GlobalScope.async {
            var result: ArticleCollectionPagerAdapter? = null
            val articleUrl = intent.data
            try {
                if (articleUrl != null) {
                    result = createFromUri(app, articleUrl)
                } else {
                    val action = intent.action
                    result = if (action == null) {
                        createFromLastResult(app)
                    } else if (action == "showBookmarks") {
                        createFromBookmarks(app)
                    } else if (action == "showHistory") {
                        createFromHistory(app)
                    } else {
                        createFromIntent(app, intent)
                    }
                }
                articleCollectionPagerAdapter = result
                if (articleCollectionPagerAdapter == null || articleCollectionPagerAdapter!!.itemCount == 0) {
                    val messageId = if (articleCollectionPagerAdapter == null) {
                        R.string.article_collection_invalid_link
                    } else {
                        R.string.article_collection_nothing_found
                    }
                    Toast.makeText(
                        this@ArticleCollectionActivity, messageId,
                        Toast.LENGTH_SHORT
                    ).show()
                    finish()
                    return@async
                }

                if (position > articleCollectionPagerAdapter!!.itemCount - 1) {
                    Toast.makeText(
                        this@ArticleCollectionActivity,
                        R.string.article_collection_selected_not_available,
                        Toast.LENGTH_SHORT
                    ).show()
                    finish()
                    return@async
                }
                withContext(Dispatchers.Main) {
                    setContentView(R.layout.activity_article_collection)

                    val tabs = findViewById<TabLayout>(R.id.tabs)
                    tabs.visibility =
                        if (articleCollectionPagerAdapter!!.itemCount == 1) ViewGroup.GONE else ViewGroup.VISIBLE

                    viewPager = findViewById<View>(R.id.pager) as ViewPager2
                    viewPager!!.isUserInputEnabled = false
                    viewPager!!.isNestedScrollingEnabled = true
                    viewPager!!.adapter = articleCollectionPagerAdapter
                    val tabLayoutMediator =
                        TabLayoutMediator(tabs, viewPager!!, true) { tab, position ->
                            tab.setText(articleCollectionPagerAdapter!!.getPageTitle(position))
                        }
                    tabLayoutMediator.attach()
                    viewPager!!.registerOnPageChangeCallback(object : OnPageChangeCallback() {
                        override fun onPageScrollStateChanged(arg0: Int) {}

                        override fun onPageScrolled(arg0: Int, arg1: Float, arg2: Int) {}

                        override fun onPageSelected(position: Int) {
                            updateTitle(position)
                            // fragment might not be created yet
                            val fragment =
                                articleCollectionPagerAdapter!!.getPageFragment(position.toLong())
                            if (fragment != null) {
                                articleCollectionPagerAdapter!!.primaryItem = fragment
                                runOnUiThread {
                                    fragment.loadUrlIfNeeded()
                                    fragment.applyTextZoomPref()
                                }
                            }
                        }
                    })


                    articleCollectionPagerAdapter!!.registerAdapterDataObserver(object :
                        AdapterDataObserver() {
                        override fun onChanged() {
                            if (articleCollectionPagerAdapter!!.itemCount == 0) {
                                finish()
                            }
                        }
                    })
                    viewPager!!.setCurrentItem(position, false)
                    updateTitle(position)
                }

            } catch (e: Exception) {
                Handler(Looper.getMainLooper()).post {
                    Toast.makeText(
                        this@ArticleCollectionActivity,
                        e!!.localizedMessage,
                        Toast.LENGTH_SHORT
                    ).show()
                    finish()
                }
            }
        }
        deferred.start()
    }

    private fun createFromUri(app: Application, articleUrl: Uri): ArticleCollectionPagerAdapter? {
        val host = articleUrl.host
        if (!(host == "localhost" || host!!.matches("127.\\d{1,3}.\\d{1,3}.\\d{1,3}".toRegex()))) {
            return createFromIntent(app, intent)
        }
        val bd = BlobDescriptor.fromUri(articleUrl) ?: return null
        val result = app.find(bd.key, bd.slobId)
        val data = BlobListAdapter(this, 20, 1)
        data.setData(result)
        val hasFragment = !Util.isBlank(bd.fragment)
        return ArticleCollectionPagerAdapter(
            app, data, if (hasFragment) ToBlobWithFragment(bd.fragment) else blobToBlob, this
        )
    }

    private fun createFromLastResult(app: Application): ArticleCollectionPagerAdapter {
        return ArticleCollectionPagerAdapter(
            app, app.lastResult, blobToBlob, this
        )
    }

    private fun createFromBookmarks(app: Application): ArticleCollectionPagerAdapter {
        return ArticleCollectionPagerAdapter(
            app, BlobDescriptorListAdapter(app.bookmarks), object : ToBlob {
                override fun convert(item: Any): Slob.Blob? {
                    return app.bookmarks!!.resolve((item as BlobDescriptor))
                }
            }, this
        )
    }

    private fun createFromHistory(app: Application): ArticleCollectionPagerAdapter {
        return ArticleCollectionPagerAdapter(
            app, BlobDescriptorListAdapter(app.history), object : ToBlob {
                override fun convert(item: Any): Slob.Blob? {
                    return app.history!!.resolve((item as BlobDescriptor))
                }
            }, this
        )
    }

    private fun createFromIntent(app: Application, intent: Intent): ArticleCollectionPagerAdapter {
        var lookupKey = intent.getStringExtra(Intent.EXTRA_TEXT)
        if (intent.action == Intent.ACTION_PROCESS_TEXT) {
            lookupKey = getIntent()
                .getCharSequenceExtra(Intent.EXTRA_PROCESS_TEXT).toString()
        }
        if (lookupKey == null) {
            lookupKey = intent.getStringExtra(SearchManager.QUERY)
        }
        if (lookupKey == null) {
            lookupKey = intent.getStringExtra("EXTRA_QUERY")
        }
        var preferredSlobId: String? = null
        if (lookupKey == null) {
            val uri = intent.data
            if (uri != null) {
                val segments = uri!!.pathSegments
                val length = segments.size
                if (length > 0) {
                    lookupKey = segments[length - 1]
                }
                val slobUri = Util.wikipediaToSlobUri(uri)
                Log.d(TAG, String.format("Converted URI %s to slob URI %s", uri, slobUri))
                if (slobUri != null) {
                    val slob = app.findSlob(slobUri)
                    if (slob != null) {
                        preferredSlobId = slob.id.toString()
                        Log.d(
                            TAG,
                            String.format("Found slob %s for slob URI %s", preferredSlobId, slobUri)
                        )
                    }
                }

            }
        }
        val data = BlobListAdapter(this, 20, 1)
        if (lookupKey == null || lookupKey.length == 0) {
            val msg = getString(R.string.article_collection_nothing_to_lookup)
            throw RuntimeException(msg)
        } else {
            val result = stemLookup(app, lookupKey, preferredSlobId)
            data.setData(result)
        }
        return ArticleCollectionPagerAdapter(
            app, data, blobToBlob, this
        )
    }

    private fun stemLookup(
        app: Application,
        lookupKey: String,
        preferredSlobId: String? = null
    ): Iterator<Slob.Blob> {
        var result: PeekableIterator<Slob.Blob>
        val length = lookupKey.length
        var currentLookupKey = lookupKey
        var currentLength = currentLookupKey.length
        do {
            result = app.find(currentLookupKey, preferredSlobId, true)
            if (result.hasNext()) {
                val b = result.peek()
                if (b.key.length - length > 3) {
                    //we don't like this result
                } else {
                    break
                }
            }
            currentLookupKey = currentLookupKey.substring(0, currentLength - 1)
            currentLength = currentLookupKey.length
        } while (length - currentLength < 5 && currentLength > 0)
        return result
    }

    private fun updateTitle(position: Int) {
        Log.d("updateTitle", "" + position + " count: " + articleCollectionPagerAdapter!!.itemCount)
        val blob = articleCollectionPagerAdapter!!.get(position)
        val pageTitle = articleCollectionPagerAdapter!!.getPageTitle(position)
        Log.d("updateTitle", "" + blob)
        val actionBar = supportActionBar
        if (blob != null) {
            val dictLabel = blob.owner.tags["label"]
            actionBar!!.title = dictLabel
            val app = application as Application
            app.history!!.add(app.getUrl(blob))
        } else {
            actionBar!!.setTitle("???")
        }
        actionBar.subtitle = pageTitle
    }


    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String?) {
        if (key == PREF_FULLSCREEN) {
            applyFullScreenPref()
        }
    }

    private fun applyFullScreenPref() {
        if (fullScreenPref) {
            fullScreen()
        } else {
            unFullScreen()
        }
    }

    fun prefs(): SharedPreferences {
        return getSharedPreferences(PREF, MODE_PRIVATE)
    }

    var fullScreenPref: Boolean
        get() = prefs().getBoolean(PREF_FULLSCREEN, false)
        private set(value) {
            val editor = prefs().edit()
            editor.putBoolean(PREF_FULLSCREEN, value)
            editor.commit()
        }

    private fun fullScreen() {
        Log.d(TAG, "[F] fullscreen")
        window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_FULLSCREEN
                or View.SYSTEM_UI_FLAG_IMMERSIVE)
        supportActionBar!!.hide()
    }

    private fun unFullScreen() {
        Log.d(TAG, "[F] unfullscreen")
        window.decorView.systemUiVisibility = 0
        supportActionBar!!.show()
    }

    fun toggleFullScreen() {
        fullScreenPref = !fullScreenPref
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "[F] Resume")
        applyFullScreenPref()
        val decorView = window.decorView
        decorView.setOnSystemUiVisibilityChangeListener(this)
        prefs().registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()
        Log.d(TAG, "[F] Pause")
        val decorView = window.decorView
        decorView.setOnSystemUiVisibilityChangeListener(null)
        prefs().unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onDestroy() {
        onDestroyCalled = true
        if (viewPager != null) {
            viewPager!!.adapter = null
        }
        if (articleCollectionPagerAdapter != null) {
            articleCollectionPagerAdapter!!.destroy()
        }
        val app = application as Application
        app.pop(this)
        super.onDestroy()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                val upIntent =
                    Intent.makeMainActivity(ComponentName(this, MainActivity::class.java))
                if (NavUtils.shouldUpRecreateTask(this, upIntent)) {
                    TaskStackBuilder.create(this)
                        .addNextIntent(upIntent).startActivities()
                    finish()
                } else {
                    // This activity is part of the application's task, so simply
                    // navigate up to the hierarchical parent activity.
                    upIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP)
                    startActivity(upIntent)
                    finish()
                }
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSystemUiVisibilityChange(visibility: Int) {
        if (isFinishing) {
            return
        }
        val decorView = window.decorView
        val uiOptions = decorView.systemUiVisibility
        val isHideNavigation =
            ((uiOptions or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION) == uiOptions)
        if (!isHideNavigation) {
            fullScreenPref = false
        }
    }

    private fun useVolumeForNav(): Boolean {
        val app = application as Application
        return app.useVolumeForNav()
    }


    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        if (event.isCanceled) {
            return true
        }
        if (articleCollectionPagerAdapter == null) {
            return false
        }
        val af = articleCollectionPagerAdapter!!.primaryItem
        if (af != null) {
            val webView = af.webView
            if (webView != null) {
                if (keyCode == KeyEvent.KEYCODE_BACK) {
                    if (webView.canGoBack()) {
                        webView.goBack()
                        return true
                    }
                }

                if (keyCode == KeyEvent.KEYCODE_VOLUME_UP) {
                    if (!useVolumeForNav()) {
                        return false
                    }
                    val scrolled = webView.pageUp(false)
                    if (!scrolled) {
                        val current = currentPosition
                        if (current > 0) {
                            viewPager!!.currentItem = current - 1
                        } else {
                            finish()
                        }
                    }
                    return true
                }
                if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
                    if (!useVolumeForNav()) {
                        return false
                    }
                    val scrolled = webView.pageDown(false)
                    if (!scrolled) {
                        val current = currentPosition
                        if (current >= 0 && current < articleCollectionPagerAdapter!!.itemCount - 1) {
                            viewPager!!.currentItem = current + 1
                        }
                    }
                    return true
                }
            }
        }
        return super.onKeyUp(keyCode, event)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_VOLUME_UP || keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
            if (!useVolumeForNav()) {
                return false
            }
            event.startTracking()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onKeyLongPress(keyCode: Int, event: KeyEvent): Boolean {
        if (!useVolumeForNav()) {
            return false
        }
        val af = articleCollectionPagerAdapter!!.primaryItem
        if (af != null) {
            val webView = af.webView!!
            if (webView == null) {
                return false
            }
            if (keyCode == KeyEvent.KEYCODE_VOLUME_UP) {
                webView.pageUp(true)
                return true
            }
            if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
                webView.pageDown(true)
                return true
            }
        }
        return super.onKeyLongPress(keyCode, event)
    }


    interface ToBlob {
        fun convert(item: Any): Slob.Blob?
    }

    inner class ArticleCollectionPagerAdapter(
        private var app: Application?,
        private var data: BaseAdapter?,
        toBlob: ToBlob,
        activity: FragmentActivity?
    ) : FragmentStateAdapter(
        activity!!
    ) {
        private val observer: DataSetObserver
        private val toBlob: ToBlob
        private var count: Int
        var primaryItem: ArticleFragment? = null

        init {
            this.count = data!!.count
            this.observer = object : DataSetObserver() {
                override fun onChanged() {
                    count = data!!.count
                    notifyDataSetChanged()
                }
            }
            data!!.registerDataSetObserver(observer)
            this.toBlob = toBlob
        }

        fun destroy() {
            data!!.unregisterDataSetObserver(observer)
            data = null
            app = null
        }

        override fun createFragment(i: Int): Fragment {
            val fragment: Fragment = ArticleFragment()

            val blob = get(i)
            if (blob != null) {
                val articleUrl = app!!.getUrl(blob)
                val args = Bundle()
                args.putString(ArticleFragment.ARG_URL, articleUrl)
                args.putInt(ArticleFragment.ARG_POSITION, i)
                fragment.arguments = args
            }
            return fragment
        }

        override fun getItemCount(): Int {
            return count
        }

        fun get(position: Int): Slob.Blob? {
            return toBlob.convert(data!!.getItem(position))
        }

        override fun getItemId(position: Int): Long {
            // we override getItemId to ensure position is the id
            // so that we can find a fragment based on a position
            return position.toLong()
        }

        fun getPageFragment(id: Long): ArticleFragment? {
            return this@ArticleCollectionActivity.supportFragmentManager.findFragmentByTag(
                "f$id"
            ) as ArticleFragment?
        }


        //        @Override
        fun getPageTitle(position: Int): CharSequence {
            if (position < data!!.count) {
                val item = data!!.getItem(position)
                if (item is BlobDescriptor) {
                    return item.key
                }
                if (item is Slob.Blob) {
                    return item.key
                }
            }
            return "???"
        } //this is needed so that fragment is properly updated
        //if underlying data changes (such as on unbookmark)
        //https://code.google.com/p/android/issues/detail?id=19001
        //        @Override
        //        public int getItemPosition(Object object) {
        //            return POSITION_NONE;
        //        }
    }

    companion object {
        private val TAG: String = ArticleCollectionActivity::class.java.simpleName

        const val PREF: String = "articleCollection"
        const val PREF_FULLSCREEN: String = "fullscreen"
    }
}
